.gitignore

```diff
@@ -17,6 +17,9 @@ yarn.lock
 /composer.lock
 
 .cache
+!/.cache/
+/.cache/**
+!/.cache/.gitkeep
```

Comment:
This ignores all `.gitkeep` files but `/.caches/.gitkeep`.
The `/.caches/` folder is needed for the `PHP Coding Standards` CI job.

Reply:
This deserves an explanatory comment inline as it's going to be really hard to find this PR comment when someone is trying to understand why the `.cache` directory is so zig-zagged here. Can we add the explanation, and include why a simpler formulation of the rules cannot provide what we need?

Is it not possible to have `PHP Coding Standards` use another directory? What is the reason for its use of `.cache` here anyway?

Reply:
The reason for zig-zagging [is the following](https://git-scm.com/docs/gitignore):
```
It is not possible to re-include a file if a parent directory of that file is excluded. 
```

> This deserves an explanatory comment inline as it's going to be really hard to find this PR comment when someone is trying to understand why the .cache directory is so zig-zagged here.

I agree. Fixed in 89a4d738e13c7d6ff51d54cfa2477eb0c713c67b

Currently, Gutenberg ignores all `.cache` folders (regardless of their location).
For compatibility reasons, these folders should still be ignored, while allowing the /.cache folder (in the root).
I haven't been able to come up with a simpler pattern that would allow that.

> if this weren't .gitkeep'd then this surprising change to .gitignore presumably would also not be necessary

Git doesn't allow to commit empty folders. If it did, I would commit an empty folder.

> Is it not possible to have PHP Coding Standards use another directory? What is the reason for its use of .cache here anyway?

Yes, it's possible, but I don't understand what's wrong with `/.cache`.

A similar Core job [uses a folder with](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/coding-standards.yml#L88) the same name -  `.cache`.

Reply:
sure but why do we need or want to commit an empty directory?

Reply:
Because PHPCS expects the directory to which to write the cache file to exist. It will not create it.

So if the ruleset dictates the the cache file should be written to the `.cache` directory, that directory _must_ exist as otherwise PHPCS will error out.
You could, of course, choose to just let the cache file be written to the system temp directory or to the project root...

Note: the PHPCS cache file itself should definitely be `gitignored`.

Reply:
my suspicion was that it was something like this, but if that's the case, that doesn't explain why it needs to be in `git`, it merely notes that the directory needs to be present when running the linter.

if the entire purpose of coercing `git` to include an empty directory in the repo is so that a linter can run in CI then it would be simpler to create the directory when running the CI script with `mkdir -p .cache` or the equivalent.

Reply:
Well, that depends. If the ruleset contains the `cache` directive, local CS runs will also expect the directory.

If the caching is only used in CI, I agree, a `mkdir` would suffice, but then again, why would the caching be limited to CI ?

Also note - in most projects which use a `.cache` directory for the PHPCS cache placement, it is generally because they want to cache those files in CI and Travis (yes, that's the completely outdated reason), didn't allow for caching individual files, only allowed for caching directories, which is why people resorted to having a `.cache` directory for this.

GH Actions, however, _does_ allow for caching individual files, so I honestly don't know why that `.cache` directory is still used, especially for projects setting this up from scratch. (I can understand people not having revisited their previous reason for setting it up when converting from Travis to GHA, but that's not the case here)